server/src/main/java/io/kroki/server/service/Plantuml.java

```diff
@@ -24,25 +24,52 @@
 
 import java.io.BufferedReader;
 import java.io.ByteArrayOutputStream;
+import java.io.File;
 import java.io.IOException;
 import java.io.StringReader;
 import java.io.UnsupportedEncodingException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 public class Plantuml implements DiagramService {
 
   private static final List<FileFormat> SUPPORTED_FORMATS = Arrays.asList(FileFormat.PNG, FileFormat.SVG, FileFormat.JPEG, FileFormat.BASE64, FileFormat.TXT, FileFormat.UTXT);
 
-  private static final Pattern INCLUDE_RX = Pattern.compile("^\\s*!include(?:url)?\\s+(.*)");
+  /**
+   * Extracts the target of the include/includeurl/includesub directive. The first and only matching group is the
+   * target.
+   * <p />
+   * We ignore any subpart/sub-document identifiers, i.e. anything after a {@code !} in the include name
+   * <p />
+   * We ignore any trailing comments on the line, i.e. anything after a {@code #} in the include line
+   * <p />
+   * Some sample patterns:
+   * <ul>
+   *   <li>{@code !include &lt;some/stdlib&gt;} includes a file from the stdlib</li>
+   *   <li>{@code !include http://some/url} includes an external resource by URL</li>
+   *   <li>{@code !include /absolute/file/path} includes a file from the filesystem</li>
+   *   <li>{@code !include search-path-file} includes a file on the "plantuml.include.path"</li>
+   *   <li>{@code !include search-path-file!2} includes the 3rd document in a file on the "plantuml.include.path"</li>
+   *   <li>{@code !includesub search-path-file!SUBID} includes the SUBID subpart of a file on the "plantuml.include.path"</li>
+   *   <li>{@code !includeurl http://some/url} deprecated include of an external resource by URL</li>
+   * </ul>
+   *
+   * @see <a href="https://plantuml.com/preprocessing">PlantUML Preprocessing</a>
+   */
+  private static final Pattern INCLUDE_RX = Pattern.compile("^\\s*!include(?:url|sub)?\\s+(.*)(?:!.*)?(?:#.*)?");
```

Comment:
Much needed documentation, thanks :+1: 
What is the last part of the regular expression with the `#` symbol? Is that when a URL contains a fragment?

Reply:
that's a comment in plantuml